### ISSUE TITLE
Discussion: A race-free version of the Take() operator

### ISSUE THREAD

**Issue by andreaskoepf** [dotnet/reactive#1414]
I would like to discuss the semantics of `Take()`: Currently this operator is implemented completely without any form of synchronization, e.g. see https://github.com/dotnet/reactive/blob/d8dbf12254c0916ee4cdc5f0f3ddd312f4fefa93/Rx.NET/Source/src/System.Reactive/Linq/Observable/Take.cs#L54-L64

This leads to severe problems when sources generate overlapping `OnNext()` calls, E.g. here a minimal repro that gets stuck most of the time after less than 100 repeat-iterations:

```
class Program
{
    static async Task Main(string[] args)
    {
        async Task Generate(int i, IObserver<int> o)
        {
            for (; ; )
            {
                o.OnNext(i);
                await Task.Yield();
            }
        }

        var s = new Subject<int>();

        // spawn to background generators...
        var tasks = Enumerable.Range(0, 2).Select(x => Generate(x, s)).ToArray();

        int i = 0;
        s.Take(1).ObserveOn(TaskPoolScheduler.Default).Repeat().Subscribe(x =>
        {
            i += 1;
            Console.WriteLine(i);
        });

        await Task.WhenAll(tasks);
    }
}
```

Of course I understand that this can be quickly fixed by inserting a `Synchronize(gate)` before `Take(1)`... I am opening this issue because I spent a whole week debugging a similar race-condition in production code which only triggered after a long time and the last thing I had on my list for potential failures was  the RX `Take()` operator. I think with minimal code changes like e.g. `Interlocked.Decrement()` this race condition could be addressed even without introducing any locking.

Maybe @bartdesmet or somebody from the core team could take a look (believe me - I had a hard time debugging this).

**Comment by akarnokd** [comment]
The error is in your code because Subject.OnNext  (or more generally any IObserver method) shouldn't be called concurrently from multiple threads. 

There is a synchronization wrapper somewhere in Rx that you can use to wrap the Subject's IObserver side.

**Comment by andreaskoepf** [comment]
@akarnokd Yes, it is correct that the subject was also in the real world example called from multiple threads. But since these errors are extremely hard to debug I would recommend from a dev perspective that either the wrong use of subjects is detected/synchronized or that operators like Take are slightly modified to become more robust. Right now the "concurrency contract" is not strictly enforced by Rx and I can easily shoot myself in the foot - something I have never experienced with Tasks for example.

**Comment by andreaskoepf** [comment]
@akarnokd thanks for pointing out that `subject.Synchronize()` exists ... probably I should use that throughout  my code-base wherever I use subjects.

**Comment by andreaskoepf** [comment]
It interesting that no overload of `Subject.Synchronize()` exists to allow passing a custom synchronization object (gate), the subject instance is always used: https://github.com/dotnet/reactive/blob/e483d14c0c090ae739626af547c72a0a30c23797/Rx.NET/Source/src/System.Reactive/Subjects/Subject.Extensions.cs#L77

**Comment by andreaskoepf** [comment]
The unprotected non-thread-safe behavior of naked `Subject<>` is clearly documented: [Using Subjects](https://docs.microsoft.com/en-us/previous-versions/dotnet/reactive-extensions/hh242970(v=vs.103)). Probably these design decisions have been made long ago and it is now way too late to discuss it.

**Comment by bartdesmet** [comment]
Yes, this goes back all the way to the start of Rx, where we initially had subjects with implicit synchronization. However, these go against the design philosophy in .NET (e.g. `ArrayList` and `List<T>` also don't synchronize by default), so you have to opt in to synchronization.

Moreover, subjects are also used in operators such as `Publish` where the (required) sequential (non-concurrent) nature of the upstream producer is the safety property that should "flow" through the internal subject, so it doesn't need to do any synchronization by itself (it just adds cost that cannot be taken away, while an opt-in to synchronization is just that: an opt-in).

Long story short, in Rx 1.0 betas, we had `Subject<T>` et al. do synchronization. In a subsequent Rx 1.0 beta, we renamed these to `SafeSubject<T>` et al. while also introducing non-synchronizing variants that became `Subject<T>` as we know them today. By the time we shipped Rx 1.0, we deleted `SafeSubject<T>` et al. in favor of a compositional approach with `Synchronize` operators.

**Comment by bartdesmet** [comment]
One more thing that was part of the design decision around this is that given that a `Subject<T>` **is an** `IObserver<T>`, it should obey by the rules that conform to the message grammar for observers (no concurrent calls). We merely provide `Synchronize` methods to let us deal with conformance if the upstream caller can't guarantee it (and such uses of `Synchronize` should only occur on the "edge" of Rx, where it interacts with the outside world).

**Comment by andreaskoepf** [comment]
@bartdesmet Thanks a lot for taking the time to address my question. After re-reading parts of the documentation it became clear to me that I just had forgotten/ignored the rules. In my real use case the OnNext() call on the subject was triggered indirectly by a member function that was called potentially from concurrently running async-methods... I have many custom operators that are created with Observable.Create() in which I everywhere use synchronization objects when async/scheduled background tasks are involved. Only for the subjects I had forgotten the rule and interestingly this was in most places no problem, e.g. everywhere I used Observable.Merge() since Merge synchronizes internally.
I have equipped some of my simple subjects with Subject.Synchronize() ... hoping that this will not introduce new deadlocks (btw I probably will create a custom version of Synchronize which allows me to hand in a specific synchronization object instead the subject instance itself).
